source/Tests/MessageContractAnalyzerFixture.cs

```diff
@@ -0,0 +1,573 @@
+        public SimpleRequest(string s, int i) 
+        {
+            RequiredString = s; 
+            RequiredInt = i;
+        }
+
+        [Required]
+        public string RequiredString { get; set; }
+
+        [Required]
+        public int RequiredInt { get; set; }
+
+        [Optional]
+        public string? OptionalString { get; set; }
+
+        [Optional]
+        public int? OptionalInt { get; set; }
+
+        [Optional]
+        public List<string> OptionalStringList { get; set; } = new(); 
+    }
+    /// <summary>response</summary>
+    public class SimpleResponse : IResponse { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source);
+        }
+
+        [Test]
+        public async Task NoDiagnosticsOnWellFormedCommand()
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>command</summary>
+    public class SimpleCommand : ICommand<SimpleCommand, SimpleResponse>
+    {
+        protected SimpleCommand() { }
+        public SimpleCommand(string requiredString) 
+        {
+            RequiredString = requiredString; 
+        }
+
+        [Required]
+        public string RequiredString { get; set; }
+
+        [Optional]
+        public string? OptionalString { get; set; }
+    }
+    /// <summary>response</summary>
+    public class SimpleResponse : IResponse { }
+
+    /// <summary>command v1</summary>
+    public class SimpleCommandV1 : ICommand<SimpleCommandV1, SimpleResponseV1> { }
+    /// <summary>response v1</summary>
+    public class SimpleResponseV1 : IResponse { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source);
+        }
+
+        [Test]
+        public async Task NoDiagnosticsOnWellFormedResponse()
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    /// <summary>response</summary>
+    public class SimpleResponse : IResponse 
+    {
+        protected SimpleResponse() { }
+        public SimpleResponse(string requiredString) 
+        {
+            RequiredString = requiredString; 
+        }
+
+        [Required]
+        public string RequiredString { get; set; }
+
+        [Optional]
+        public string? OptionalString { get; set; }
+    }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source);
+        }
+
+        [Test]
+        public async Task NoDiagnosticsOnWellFormedEvent()
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    public class SimpleEvent : IEvent { }
+    public class SimpleEventV1 : IEvent { }
+}");
+
+            await Verify.VerifyAnalyzerAsync(source);
+        }
+
+        [Test]
+        public async Task EventTypesMustBeNamedCorrectly()
+        {
+            var source = WithOctopusTypes(@"
+namespace Octopus.Core.Features.ServerTasks.MessageContracts
+{
+    public class {|#0:EventWhichIsSimple|}: IEvent { }
```

Comment:
What goes on behind the scenes is that an analyzer reports an error at a source location.

A source location is 4 ints; `(StartLine, StartColumn, EndLine, EndColumn)` and so if you want to assert that your analyzer is reporting correctly you need to check those 4 values.

Microsoft's analyzer testing-library has some special syntax wherein they wrap this; If you write `{|#1  ... |}` then the library will work out the right lines and columns, and represent them with a placeholder location `1`. You can then write `WithLocation(1)` rather than `WithLocation(35, 12, 25, 29)` and it does the right thing. It's very nice.

https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix